packages/electron/src/electron-main/electron-application.ts

```diff
@@ -0,0 +1,441 @@
+    protected electronSecurityToken: ElectronSecurityToken;
+
+    @inject(ProcessArgv)
+    protected processArgv: ProcessArgv;
+
+    protected readonly electronStore = new Storage();
+
+    protected config: FrontendApplicationConfig;
+    readonly backendPort = new Deferred<number>();
+
+    async start(config: FrontendApplicationConfig): Promise<void> {
+        this.config = config;
+        this.hookApplicationEvents();
+        const port = await this.startBackend();
+        this.backendPort.resolve(port);
+        await app.whenReady();
+        await this.attachElectronSecurityToken(port);
+        await this.startContributions();
+        await this.launch({
+            secondInstance: false,
+            argv: this.processArgv.processArgvWithoutBin,
+            cwd: process.cwd()
+        });
+    }
+
+    async launch(params: ExecutionParams): Promise<void> {
+        createYargs(params.argv, params.cwd)
+            .command('$0 [<file>]', false,
+                cmd => cmd
+                    .positional('file', { type: 'string' }),
+                args => this.handleMainCommand(params, { file: args.file }),
+            ).parse();
+    }
+
+    /**
+     * Use this rather than creating `BrowserWindow` instances from scratch, since some security parameters need to be set, this method will do it.
+     *
+     * @param options
+     */
+    async createWindow(options: BrowserWindowConstructorOptions): Promise<BrowserWindow> {
+        const electronWindow = new BrowserWindow(options);
+        this.attachReadyToShow(electronWindow);
+        this.attachWebContentsNewWindow(electronWindow);
+        this.attachSaveWindowState(electronWindow);
+        this.attachWillPreventUnload(electronWindow);
+        this.attachGlobalShortcuts(electronWindow);
+        return electronWindow;
+    }
+
+    async openDefaultWindow(): Promise<BrowserWindow> {
+        const uri = await this.createWindowUri();
+        const electronWindow = await this.createWindow(this.getBrowserWindowOptions());
+        electronWindow.loadURL(uri.toString(true));
+        return electronWindow;
+    }
+
+    async openWindowWithWorkspace(workspace: string): Promise<BrowserWindow> {
+        const uri = (await this.createWindowUri()).withFragment(workspace);
+        const electronWindow = await this.createWindow(this.getBrowserWindowOptions());
+        electronWindow.loadURL(uri.toString(true));
+        return electronWindow;
+    }
+
+    /**
+     * "Gently" close all windows, application will not stop if a `beforeunload` handler returns `false`.
+     */
+    requestStop(): void {
+        app.quit();
+    }
+
+    protected async handleMainCommand(params: ExecutionParams, options: MainCommandOptions): Promise<void> {
+        if (typeof options.file === 'undefined') {
+            await this.openDefaultWindow();
+        } else {
+            await this.openWindowWithWorkspace(realpathSync(path.resolve(params.cwd, options.file)));
```

Comment:
`realpathSync` can fail here when `options.file` doesn't exists, this is my bad.

Tell me if you want me to fix this, or if you'll do it. I don't want to step on your toes.

Reply:
Go ahead, please. It's your PR eventually, I just wanted to verify the functionality by adding a sample.

Reply:
>It's your PR eventually

I hope you have taken it as a compliment ;)

Reply:
Ok I'll force push again then.